nodeJS/express/views.md

```diff
@@ -78,10 +78,41 @@ Start the server and go to the `/` route in the browser. You should see:
 EJS rocks!
 ```
 
-If you inspect the HTML in the browser's dev tools, you can see HTML is structures exactly like how we wrote the EJS template with the `message` variable replaced with its value.
+If you inspect the HTML in the browser's dev tools, you can see the HTML is structured exactly like how we wrote the EJS template with the `message` variable replaced with its value.
 
 When you hit the `/` route, `res.render("index", { message: "EJS rocks!" });` is the line that sends back the response. Since we've already defined the `views` and `view engine` app properties, the first argument of `res.render` is programmed to look for "a template called index in the specified folder", while the second argument is an object of variables that are to be made available to that specific template.
 
+### The locals variable in EJS
+
+In the example above, how did the template file know about the `message` variable? The answer is that we passed a `locals` object as the second argument to `res.render()`, which contained a `message` property. This makes it available for template engines to access local variables. We could have also set local variables onto the [`res.locals`](https://expressjs.com/en/5x/api.html#res.locals) object in our route, like so:
+
+```javascript
+app.get("/", (req, res) => {
+  res.locals.message = "EJS rocks!";
+  res.render("index");
+});
+```
+
+Then, EJS stores this in an object which is named `locals` by default (check the localsName bullet point in the Options section in the [EJS docs](https://ejs.co/#docs) to learn more). Thus the `message` variable is accessible with either `locals.message` or simply `message`.
+
+<div class="lesson-note lesson-note--tip" markdown="1">
+
+#### Undefined variables in locals
+
+If we try to access a variable in a rendered template file that was not defined in the `locals` argument of `res.render` or `res.locals`, this can cause a reference error. For instance if we try to access an undefined `foo` variable, `locals.foo` will return undefined, while `foo` will result in an reference error. Verify this by adding `locals.foo`, and then `foo` below `message` in `index.ejs`:
+
+```ejs
+<html>
+  <body>
+    <%= message %>
+    <%= locals.foo %>
+    <%= foo %>
+  </body>
+</html>
+```
```

Comment:
Non-blocking nit - what do you think? Just because with the original, it won't show the output of `message` and `locals.foo`, since it'll just show the error page for `foo`'s reference error straight away.
````suggestion
If we try to access a variable in a rendered template file that was not defined in the `locals` argument of `res.render` or `res.locals`, this can cause a reference error. For instance if we try to access an undefined `foo` variable, `locals.foo` will return undefined, while `foo` will result in an reference error. Verify this by outputting `locals.foo` in `index.ejs`, then replacing it with `foo`:

```ejs
<html>
  <body>
    <%= message %>
    <!-- replace the below with the output of just foo -->
    <%= locals.foo %>
  </body>
</html>
```
````

Reply:
This is good, your way conveys what I wanted the learner to do much more effectively.